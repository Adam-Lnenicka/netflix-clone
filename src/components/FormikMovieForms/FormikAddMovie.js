import React, { useState } from "react";
import { Formik, Form, Field } from "formik";
import * as Yup from "yup";
import { useDispatch } from "react-redux";
import {
  addMovieActionCreator,
  resetMovieActionCreator,
} from "../../store/actionCreators";
import { Link } from "react-router-dom";

const MovieSchema = Yup.object().shape({
  title: Yup.string()
    .max(50, "The entry seems to be too long")
    .required("This field is required"),

  poster_path: Yup.string().required("This field is required"),

  overview: Yup.string()
    .min(20, "The overview needs to have at least 20 characters")
    .required("This field is required"),

  release_date: Yup.string().required("This field is required"),
  genres: Yup.string().required("This field is required"),
  runtime: Yup.number().required("This field is required"),
});

const FormikAddMovie = () => {
  const [submitted, setSubmitted] = useState(false);
  const dispatch = useDispatch();

  const handleSubmit = (values) => {
    setSubmitted(true);
    values.genres = [values.genres];
    values.id = Math.random().toString();
    console.log(values);

    dispatch(addMovieActionCreator(values));
  };

  const handleReset = (values) => {
    values.genres = [values.genres];
    values.title = "";
  };

  return (
    <div className="form">
      <div className="exit">
        <Link to="/">
          <button className="exit-button">X</button>
        </Link>
      </div>
      <h1>Add Movie</h1>

      <Formik
        initialValues={{
          title: "",
          release_date: "",
          poster_path: "",
          genres: [],
          overview: "",
          runtime: "",
        }}
        validationSchema={MovieSchema}
        onSubmit={handleSubmit}
      >
        {({ errors, touched }) => (
          <Form className="form__content">
            <div className="form__contact-field-box">
              <div>
                <label htmlFor="title">Title</label>
                <br />

                <Field
                  name="title"
                  className="form__input"
                  placeholder="Title"
                  type="text"
                  id="title"
                />
                {errors.title && touched.title ? (
                  <div className="form__error-message">{errors.title}</div>
                ) : null}
              </div>
              <label htmlFor="release_date">Release Date</label>
              <br />

              <Field
                name="release_date"
                placeholder="Choose Date"
                className="form__input"
                type="text"
                id="release_date"
              />
              {errors.release_date && touched.release_date ? (
                <div className="form__error-message">{errors.release_date}</div>
              ) : null}

              <div>
                <label htmlFor="poster_path">Movie URL</label>
                <br />

                <Field
                  name="poster_path"
                  placeholder="Movie URL here"
                  className="form__input"
                  type="text"
                  id="poster_path"
                />
                {errors.poster_path && touched.poster_path ? (
                  <div className="form__error-message">
                    {errors.poster_path}
                  </div>
                ) : null}
              </div>
              <label htmlFor="genres">Genre</label>
              <br />

              <Field
                name="genres"
                placeholder="Genre"
                className="form__input"
                type="text"
                id="genres"
              />
              {errors.genres && touched.genres ? (
                <div className="form__error-message">{errors.genres}</div>
              ) : null}
              <div>
                <label htmlFor="overview">Overview</label>
                <br />

                <Field
                  name="overview"
                  placeholder="Overview"
                  className="form__input"
                  type="text"
                  id="overview"
                />
                {errors.overview && touched.overview ? (
                  <div className="form__error-message">{errors.overview}</div>
                ) : null}
              </div>
              <div>
                <label htmlFor="runtime">Runtime</label>
                <br />

                <Field
                  name="runtime"
                  placeholder="Runtime"
                  className="form__input"
                  type="text"
                  id="runtime"
                />
                {errors.runtime && touched.runtime ? (
                  <div className="form__error-message">{errors.runtime}</div>
                ) : null}
              </div>
            </div>

            {errors.notes && touched.notes ? (
              <div className="form__error-message">{errors.notes}</div>
            ) : null}
            <div className="button-area">
              <button className="button-secondary" type="submit">
                reset
              </button>{" "}
              <button className="button-main" type="submit">
                Submit
              </button>
            </div>
          </Form>
        )}
      </Formik>
      <p>
        {submitted ? (
          <div className="form__submit-message">Movie will be added</div>
        ) : null}
      </p>
    </div>
  );
};

export default FormikAddMovie;
