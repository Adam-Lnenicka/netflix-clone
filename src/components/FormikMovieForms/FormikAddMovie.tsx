import React, { useState } from "react";
import { Formik, Form } from "formik";
import * as Yup from "yup";
import {
  addMovieActionCreator,
  resetMovieActionCreator,
} from "../../store/actionCreators";
import { Link } from "react-router-dom";
import FormikField from "./FormikField";
import { fieldArray } from "./fieldArray";
import { useAppDispatch } from "../../store/selectors";

const MovieSchema = Yup.object().shape({
  title: Yup.string()
    .max(50, "The entry seems to be too long")
    .required("This field is required"),

  poster_path: Yup.string().required("This field is required"),

  overview: Yup.string()
    .min(20, "The overview needs to have at least 20 characters")
    .required("This field is required"),

  release_date: Yup.string().required("This field is required"),
  genres: Yup.string().required("This field is required"),
  runtime: Yup.number().required("This field is required"),
});

const FormikAddMovie: React.FC = () => {
  const [submitted, setSubmitted] = useState(false);
  const dispatch = useAppDispatch();

  const handleSubmit = (values) => {
    setSubmitted(true);
    values.genres = [values.genres];
    values.id = Math.random().toString();
    console.log(values);

    dispatch(addMovieActionCreator(values));
  };

  const handleReset = (values) => {
    values.genres = [values.genres];
    values.title = "";
  };

  interface AddMovieInterface {
    title: string;
    release_date: string;
    poster_path: string;
    genres: string[];
    overview: string;
    runtime: string;
  }

  const initialValues: AddMovieInterface = {
    title: "",
    release_date: "",
    poster_path: "",
    genres: [],
    overview: "",
    runtime: "",
  };

  return (
    <div className="form">
      <div className="exit">
        <Link to="/">
          <button className="exit-button">X</button>
        </Link>
      </div>
      <h1>Add Movie</h1>

      <Formik
        initialValues={initialValues}
        validationSchema={MovieSchema}
        onSubmit={handleSubmit}
      >
        {({ errors, touched }) => (
          <Form className="form__content">
            <div className="form__contact-field-box">
              {fieldArray.map((array) => {
                return (
                  <FormikField
                    key={array.title}
                    title={array.title}
                    path={array.path}
                    id={array.path}
                    myErrors={errors}
                    myTouched={touched}
                  />
                );
              })}
              <button className="button-secondary" type="submit">
                reset
              </button>{" "}
              <button className="button-main" type="submit">
                Submit
              </button>
            </div>
          </Form>
        )}
      </Formik>
      <p>
        {submitted ? (
          <div className="form__submit-message">Movie will be added</div>
        ) : null}
      </p>
    </div>
  );
};

export default FormikAddMovie;
