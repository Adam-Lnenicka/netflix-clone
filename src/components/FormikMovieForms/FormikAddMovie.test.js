import "jsdom-global/register";

import React from "react";
import { mount, shallow } from "enzyme";
import FormikAddMovie from "./FormikAddMovie";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import toJson from "enzyme-to-json";

describe("FormikAddMovie", () => {
  const initialState = {};
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const onSubmitFn = jest.fn();

  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <FormikAddMovie onSubmit={onSubmitFn} />
      </Provider>
    );
  });

  it("calls onSubmit when add formik form submitted", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("calls onSubmit when add formik form submitted", () => {
    const submitButton = wrapper.find(".button-main").simulate("click");
    expect(submitButton.exists()).toBe(true);
  });
});
