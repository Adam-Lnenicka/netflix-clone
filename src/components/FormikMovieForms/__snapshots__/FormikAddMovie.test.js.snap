// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormikAddMovie calls onSubmit when add formik form submitted 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <FormikAddMovie
    onSubmit={[MockFunction]}
  >
    <div
      className="form"
    >
      <div
        className="exit"
      />
      <h1>
        Add Movie
      </h1>
      <Formik
        initialValues={
          Object {
            "genres": Array [],
            "overview": "",
            "poster_path": "",
            "release_date": "",
            "runtime": "",
            "title": "",
          }
        }
        onSubmit={[Function]}
        validationSchema={
          ObjectSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_excludedEdges": Array [],
            "_nodes": Array [
              "runtime",
              "genres",
              "release_date",
              "overview",
              "poster_path",
              "title",
            ],
            "_sortErrors": [Function],
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {},
            "fields": Object {
              "genres": StringSchema {
                "_blacklist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_blacklistError": undefined,
                "_mutate": undefined,
                "_typeError": [Function],
                "_whitelist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_whitelistError": undefined,
                "conditions": Array [],
                "deps": Array [],
                "exclusiveTests": Object {
                  "required": true,
                },
                "spec": Object {
                  "abortEarly": true,
                  "nullable": false,
                  "presence": "required",
                  "recursive": true,
                  "strict": false,
                  "strip": false,
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
                "type": "string",
              },
              "overview": StringSchema {
                "_blacklist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_blacklistError": undefined,
                "_mutate": undefined,
                "_typeError": [Function],
                "_whitelist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_whitelistError": undefined,
                "conditions": Array [],
                "deps": Array [],
                "exclusiveTests": Object {
                  "min": true,
                  "required": true,
                },
                "spec": Object {
                  "abortEarly": true,
                  "nullable": false,
                  "presence": "required",
                  "recursive": true,
                  "strict": false,
                  "strip": false,
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
                "type": "string",
              },
              "poster_path": StringSchema {
                "_blacklist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_blacklistError": undefined,
                "_mutate": undefined,
                "_typeError": [Function],
                "_whitelist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_whitelistError": undefined,
                "conditions": Array [],
                "deps": Array [],
                "exclusiveTests": Object {
                  "required": true,
                },
                "spec": Object {
                  "abortEarly": true,
                  "nullable": false,
                  "presence": "required",
                  "recursive": true,
                  "strict": false,
                  "strip": false,
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
                "type": "string",
              },
              "release_date": StringSchema {
                "_blacklist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_blacklistError": undefined,
                "_mutate": undefined,
                "_typeError": [Function],
                "_whitelist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_whitelistError": undefined,
                "conditions": Array [],
                "deps": Array [],
                "exclusiveTests": Object {
                  "required": true,
                },
                "spec": Object {
                  "abortEarly": true,
                  "nullable": false,
                  "presence": "required",
                  "recursive": true,
                  "strict": false,
                  "strip": false,
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
                "type": "string",
              },
              "runtime": NumberSchema {
                "_blacklist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_blacklistError": undefined,
                "_mutate": undefined,
                "_typeError": [Function],
                "_whitelist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_whitelistError": undefined,
                "conditions": Array [],
                "deps": Array [],
                "exclusiveTests": Object {
                  "required": true,
                },
                "spec": Object {
                  "abortEarly": true,
                  "nullable": false,
                  "presence": "required",
                  "recursive": true,
                  "strict": false,
                  "strip": false,
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
                "type": "number",
              },
              "title": StringSchema {
                "_blacklist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_blacklistError": undefined,
                "_mutate": undefined,
                "_typeError": [Function],
                "_whitelist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_whitelistError": undefined,
                "conditions": Array [],
                "deps": Array [],
                "exclusiveTests": Object {
                  "max": true,
                  "required": true,
                },
                "spec": Object {
                  "abortEarly": true,
                  "nullable": false,
                  "presence": "required",
                  "recursive": true,
                  "strict": false,
                  "strip": false,
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
                "type": "string",
              },
            },
            "spec": Object {
              "abortEarly": true,
              "nullable": false,
              "presence": "optional",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
            "type": "object",
          }
        }
      >
        <Form
          className="form__content"
        >
          <form
            action="#"
            className="form__content"
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <div
              className="form__contact-field-box"
            >
              <FormikField
                id="title"
                key="Title"
                myErrors={Object {}}
                myTouched={Object {}}
                path="title"
                title="Title"
              >
                <div>
                  <div
                    className="form__label"
                  >
                    Title
                  </div>
                  <Field
                    className="form__input"
                    id="title"
                    name="title"
                    placeholder="Title"
                    type="text"
                  >
                    <input
                      className="form__input"
                      id="title"
                      name="title"
                      onBlur={[Function]}
                      onChange={[Function]}
                      placeholder="Title"
                      type="text"
                      value=""
                    />
                  </Field>
                </div>
              </FormikField>
              <FormikField
                id="release_date"
                key="Release Date"
                myErrors={Object {}}
                myTouched={Object {}}
                path="release_date"
                title="Release Date"
              >
                <div>
                  <div
                    className="form__label"
                  >
                    Release Date
                  </div>
                  <Field
                    className="form__input"
                    id="release_date"
                    name="release_date"
                    placeholder="Release Date"
                    type="text"
                  >
                    <input
                      className="form__input"
                      id="release_date"
                      name="release_date"
                      onBlur={[Function]}
                      onChange={[Function]}
                      placeholder="Release Date"
                      type="text"
                      value=""
                    />
                  </Field>
                </div>
              </FormikField>
              <FormikField
                id="poster_path"
                key="Poster Path"
                myErrors={Object {}}
                myTouched={Object {}}
                path="poster_path"
                title="Poster Path"
              >
                <div>
                  <div
                    className="form__label"
                  >
                    Poster Path
                  </div>
                  <Field
                    className="form__input"
                    id="poster_path"
                    name="poster_path"
                    placeholder="Poster Path"
                    type="text"
                  >
                    <input
                      className="form__input"
                      id="poster_path"
                      name="poster_path"
                      onBlur={[Function]}
                      onChange={[Function]}
                      placeholder="Poster Path"
                      type="text"
                      value=""
                    />
                  </Field>
                </div>
              </FormikField>
              <FormikField
                id="genres"
                key="Genre"
                myErrors={Object {}}
                myTouched={Object {}}
                path="genres"
                title="Genre"
              >
                <div>
                  <div
                    className="form__label"
                  >
                    Genre
                  </div>
                  <Field
                    className="form__input"
                    id="genres"
                    name="genres"
                    placeholder="Genre"
                    type="text"
                  >
                    <input
                      className="form__input"
                      id="genres"
                      name="genres"
                      onBlur={[Function]}
                      onChange={[Function]}
                      placeholder="Genre"
                      type="text"
                      value={Array []}
                    />
                  </Field>
                </div>
              </FormikField>
              <FormikField
                id="overview"
                key="Overview"
                myErrors={Object {}}
                myTouched={Object {}}
                path="overview"
                title="Overview"
              >
                <div>
                  <div
                    className="form__label"
                  >
                    Overview
                  </div>
                  <Field
                    className="form__input"
                    id="overview"
                    name="overview"
                    placeholder="Overview"
                    type="text"
                  >
                    <input
                      className="form__input"
                      id="overview"
                      name="overview"
                      onBlur={[Function]}
                      onChange={[Function]}
                      placeholder="Overview"
                      type="text"
                      value=""
                    />
                  </Field>
                </div>
              </FormikField>
              <FormikField
                id="runtime"
                key="Runtime"
                myErrors={Object {}}
                myTouched={Object {}}
                path="runtime"
                title="Runtime"
              >
                <div>
                  <div
                    className="form__label"
                  >
                    Runtime
                  </div>
                  <Field
                    className="form__input"
                    id="runtime"
                    name="runtime"
                    placeholder="Runtime"
                    type="text"
                  >
                    <input
                      className="form__input"
                      id="runtime"
                      name="runtime"
                      onBlur={[Function]}
                      onChange={[Function]}
                      placeholder="Runtime"
                      type="text"
                      value=""
                    />
                  </Field>
                </div>
              </FormikField>
              <button
                className="button-secondary"
                type="submit"
              >
                reset
              </button>
               
              <button
                className="button-main"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </Form>
      </Formik>
      <p />
    </div>
  </FormikAddMovie>
</Provider>
`;
